### PR TITLE
Executor need reasonable default value

### LINKED ISSUE #5648: Executor need reasonable default value
Currently Executor needs the place argument. It could be improved by making it optional, PaddlePaddle can choose the reasonable default value: if the user has GPU, choose GPU, otherwise choose CPU.

### PR DESCRIPTION
**Pull request by tonyyang-svail**
fix #5648 

When using `exe = Executor(place=None)`, the logic is the following:

|                 | compiled without GPU | compiled with GPU     |
|-----------------|----------------------|-----------------------|
| run without GPU | CPU                  | warning, then use CPU |
| run with GPU    | CPU                  | GPU                   |

### CODE REVIEW

File: python/paddle/v2/fluid/executor.py [hunk 1]
```diff
@@ -5,7 +5,13 @@
 
 
 class Executor(object):
-    def __init__(self, places):
+    def __init__(self, places=None):
+        if places is None:
+            if core.is_compile_gpu():
```
Comment (helinwang) [inline python/paddle/v2/fluid/executor.py]: I think we need to detect at runtime whether a GPU is present. It's very likely a compiled GPU version runs on a CPU only machine.

File: paddle/pybind/pybind.cc [hunk 2]
```diff
@@ -78,6 +78,14 @@ bool IsCompileGPU() {
 #endif
 }
 
+int GetCUDADeviceCount() {
+#ifndef PADDLE_WITH_CUDA
+  return 0;
```
Comment (helinwang) [inline paddle/pybind/pybind.cc]: Maybe return -1 here to indicate error.

File: paddle/pybind/pybind.cc [hunk 3]
```diff
@@ -497,6 +505,7 @@ All parameter, weight, gradient are variables in Paddle.
   m.def("init_gflags", InitGflags);
 
   m.def("is_compile_gpu", IsCompileGPU);
+  m.def("get_cuda_device_count", GetCUDADeviceCount());
```
Comment (helinwang) [inline paddle/pybind/pybind.cc]: Not familiar with Pybind, why it's a function invocation here? `GetCUDADeviceCount()`, rather than `GetCUDADeviceCount`?

Comment (tonyyang-svail) [reply paddle/pybind/pybind.cc]: It is a bug. Thanks for pointing it out.

**Comment by helinwang** [comment]
Thanks for the logic table! Very helpful.

**Comment by tonyyang-svail** [comment]
@reyoung Thanks for your comments. The reason for this PR is more of python API thing. We want to hide `core.CPUPlace` and `core.GPUPace` from the user. We can achieve this by either have a default value (like the current PR) or use the interface like `Executor(is_gpu=False)`.

**Pull request closed without merging**